src/main/java/com/tealium/CollectDispatcher.java

```diff
@@ -0,0 +1,250 @@
+package com.tealium;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.tealium.Tealium.DispatchCallback;
+import com.tealium.DataManager.InfoKey;
+
+
+/**
+ * Tealium Collect Dispatch service for delivering tracked data to Tealium's
+ * Collect Service
+ *
+ * @author Jason Koo, Chad Hartman, Karen Tamayo, Merritt Tidwell, Chris Anderberg
+ */
+final class CollectDispatcher {
+
+    public final static String DEFAULT_URL = "https://collect.tealiumiq.com/event";
+
+    private final String endpoint;
+    private final Logger logger;
+    private final int timeout;
+
+    // =========================================================================
+    // PUBLIC
+    // =========================================================================
+
+    /**
+     * Constructor for creating an instance of the Tealium Collect Dispatch
+     * service
+     * 
+     * @param endpoint
+     *            The target url to send track dispatches to. Should NOT include
+     *            any ending query strings or "?" suffix.
+     */
+    public CollectDispatcher(String endpoint, LibraryContext context, int timeout) {
+        super();
+        this.endpoint = endpoint;
+        this.logger = context.getLogger();
+        this.timeout = timeout;
+    }
+
+    /**
+     * Packages data sources into expected URL call format and sends.
+     * 
+     * @param data
+     *            Map of all key-values to be sent with dispatch.
+     * @param callback
+     *            Optional callback object implementing the CollectCallback
+     *            interface.
+     * @throws CollectDispatchException
+     * @see{@link #CollectCallback}
+     */
+    public void dispatch(Udo data, DispatchCallback callback) throws CollectDispatchException {
```

Comment:
There's a number of try/catch blocks in this function. Seems a little busy and potentially bug prone(?). I would suggest breaking this up into small methods -  .createConnection(), .getResponseCode(), .sendData(), etc.

Reply:
Instead of throwing generic exceptions like "Exception" and "IOException", I created more specific exceptions with more semantic names, such as "CollectDispatchException". In cases where IOException was thrown in multiple places for different reasons, this could have resulted in a single try/catch block in calling code. However there are really different causes for IOException, and giving a more specific exception for each cause requires handling each for an exception separately, which is what Java's checked exceptions are really going for. The extra try/catch blocks you see are throwing more specific exceptions with more descriptive messages. In regards to bugs, I never actually throw a new exception in any new places, it's all about catching and re throwing existing exceptions, so Java's checked exception system would catch any errors should I change an exception type in one place but not another, or have forgotten to throw the new exception type in a method signature.

In regards to all the exception cases, I agree it could be buggy, but I haven't actually introduced any new exception cases, just renamed them or made them more specific. Fundamentally the architecture is unchanged with this update. I had a conversation with Chad after the code review about how we can improve the architecture around sending network requests. See [this card](https://trello.com/c/gPBUFkel/453-java-lib-collect-call-refactor) for details.